business/iter8.go

```diff
@@ -0,0 +1,279 @@
+package business
+
+import (
+	"encoding/json"
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kube "k8s.io/client-go/kubernetes"
+	"github.com/kiali/kiali/config"
+	"k8s.io/client-go/rest"
+	"time"
+	"fmt"
+	"github.com/kiali/kiali/log"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	iter8v1alpha1 "github.com/iter8-tools/iter8-controller/pkg/apis/iter8/v1alpha1"
+	kialiConfig "github.com/kiali/kiali/config"
+)
+
+const GroupName = "iter8.tools"
+const GroupVersion = "v1alpha1"
+
+type Iter8Service struct {
+	k8s kubernetes.IstioClientInterface
+	businessLayer *Layer
+}
+
+func (in *Iter8Service) GetIter8Info() (models.Iter8Info, error) {
+
+	var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: GroupVersion}
+	var iter8Kind = SchemeGroupVersion.WithKind("Experiment")
+
+	if (iter8Kind == schema.GroupVersionKind{}) {
+		return models.Iter8Info{}, nil
+	}
+	log.Infof("Experiment CRD exists")
+	return models.Iter8Info{
+		Enabled: true,
+	}, nil
+}
+
+func (in *Iter8Service) GetIter8Experiment(namespace string, name string) (experiment models.ExperimentDetail, err error) {
+	result := &iter8v1alpha1.Experiment{}
+	cacheToken := ""
+	kConfig := kialiConfig.Get()
+	if kConfig.InCluster {
+		if saToken, err := kubernetes.GetKialiToken(); err != nil {
+			return models.ExperimentDetail{}, err
+		} else {
+			cacheToken = saToken
+		}
+	}
+	config, err := kubernetes.ConfigClient()
+	if err != nil {
+		return models.ExperimentDetail{}, err
+	}
+	istioConfig := rest.Config{
+		Host:            config.Host,
+		TLSClientConfig: config.TLSClientConfig,
+		QPS:             config.QPS,
+		BearerToken:     cacheToken,
+		Burst:           config.Burst,
+	}
+
+	istioClient, err := kubernetes.NewClientFromConfig(&istioConfig)
+	clientset := istioClient.GetK8sApi()
+	path := fmt.Sprintf("/apis/iter8.tools/v1alpha1/")
+	err = clientset.CoreV1().RESTClient().Get().AbsPath(path).
+		Namespace(namespace).
+		Resource("experiments").
+		Name(name).
+		Do().Into(result)
+
+	criterias := make([]models.CriteriaDetail, len(result.Spec.Analysis.SuccessCriteria))
+	for i, c := range result.Spec.Analysis.SuccessCriteria {
```

Comment:
Normally the parsing between the raw result from a REST call and Kiali model is done inside the model/* types, trying to add a Parse() set of methods (for collections and items, i.e. model/virtual_service.go or similar). 

On this case we see a new scenario and we need to decide what to do, I tend to think that following the Kiali convetion will help, even if that means that perhaps that it needs to re-organize a little bit the existing code.

Hopefully it's a basic call & parse scenario and I think it shouldn't be a big effort on this.

I can also to try something locally and send a PR to this branch as well, I knew that probably we will face new scenarios like importing a new API ("/apis/iter.tools/v1alpha1") and in the Extensions documentations I didn't detail this.

Reply:
Yes, I can re-organize the existing code to confirm to Kiali model.